Documents the release process for Pyrsia

betarelease opened the pull request:
## Description

Starts work on pyrsia/pyrsia #1159 

This PR documents the release process so that we can use this as a reference and keep updating it as we release and add features.


## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/community/get_involved/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).

ihcomega56 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,27 @@
+
+## Before tagging the release in github
+
+- [] Make sure you update the version in `Cargo.toml`. Run the PR through the github actions to ensure nothing fails.
```
A white space is needed between brackets. It is not treated as a checkbox now. 
![image](https://user-images.githubusercontent.com/8938191/207529274-65394190-f6ea-462c-8ebd-3273eab93bf7.png)

ihcomega56 replied:
Also we need to update GitHub Actions itself? According to the previous release - https://github.com/pyrsia/pyrsia/pull/1349/files

ihcomega56 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,27 @@
+- [] Commit the PR to the main branch
+- [] Once all the above steps are completed and verified to be success, tag the release in github. Github also helps you add the release notes - although they could use some context and explaning to make them easy to understand. Fill in that text to simplify the release notes and add the generated release notes in the release.
+
+## After Tagging the release
```
Is this kind of blog post one of tasks to be listed here? https://pyrsia.io/blog/2022/11/30/pyrsia-0.2.1-released/

betarelease replied:
indeed. Will add it to the list.

ihcomega56 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,27 @@
+- [] Run integration tests and ensure there are no failures - also ensure there no flakiness observed.
+- [] Record the results of the integration tests so that we can attach them to the release notes.
+- [] Run manual confidence tests - this requires to run Pyrsia node as part of the network and exercising the new features manually. We may be able to automate parts of this but not completely since the network itself is complex to build a workflow on.
+- [] verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
```
It is better that all sentences start with a capital letter for consistency.

fishseabowl replied:
```suggestion
- [] Verify that the rust toolchain version is set to the version we would like to release. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
```

fishseabowl commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,27 @@
+
+- [] Make sure you update the version in `Cargo.toml`. Run the PR through the github actions to ensure nothing fails.
+This is the PR that will be tagged with the version number that matches the one in the `Cargo.toml`
+- [] Run integration tests and ensure there are no failures - also ensure there no flakiness observed.
```
```suggestion
- [] Run integration tests and ensure there are no failures - also ensure there is no flakiness observed.
```

fishseabowl commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,27 @@
+## Tag the release
+
+- [] Commit the PR to the main branch
+- [] Once all the above steps are completed and verified to be success, tag the release in github. Github also helps you add the release notes - although they could use some context and explaning to make them easy to understand. Fill in that text to simplify the release notes and add the generated release notes in the release.
```
small typo
```suggestion
- [] Once all the above steps are completed and verified to be successful, tag the release in github. Github also helps you add the release notes - although they could use some context and explaining to make them easy to understand. Fill in that text to simplify the release notes and add the generated release notes in the release.
```

betarelease commented:
@erwin1 Once this is complete - please link to your test plan doc here.

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+- [ ] Record the results of the integration tests so that we can attach them to the release notes.
+- [ ] Run manual confidence tests - this requires to run Pyrsia node as part of the network and exercising the new features manually. We may be able to automate parts of this but not completely since the network itself is complex to build a workflow on.
+- [ ] Verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
+- [ ] Make sure apt repo and brew repo contain the correct latest release
```
this should go under Tag the release

betarelease replied:
for `Tag the release` I kept only the steps that actually apply the tag. So moving items about running the tests as part of `After tagging`

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+- [ ] Run manual confidence tests - this requires to run Pyrsia node as part of the network and exercising the new features manually. We may be able to automate parts of this but not completely since the network itself is complex to build a workflow on.
+- [ ] Verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
+- [ ] Make sure apt repo and brew repo contain the correct latest release
+- [ ] Upload windows MSI to github release
```
this can only be done after tagging/building. move to 'Tag the release'

betarelease replied:
created new headers for sections that are for testing and deployment.

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+- [ ] Verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
+- [ ] Make sure apt repo and brew repo contain the correct latest release
+- [ ] Upload windows MSI to github release
+- [ ] Deploy the authorized nodes with this release
```
This can't be done before tagging. But I think this is a duplicate of `Deploy to production` which we already have under 'After tagging the release`

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+- [ ] Upload windows MSI to github release
+- [ ] Deploy the authorized nodes with this release
+- [ ] Run installer tests connecting to nightly
+- [ ] Update documentation to include latest released version
```
Move to 'tag the release' or maybe even better to 'after tagging' - I think it's one the final steps

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+## Tag the release
+
+- [ ] Commit the PR to the main branch
+- [ ] Update github actions with the new version number eg. <https://github.com/pyrsia/pyrsia/pull/1349/files>
```
this must be part of the very first bullet: Make sure you update the version in `Cargo.toml`

erwin1 commented on docs/developers/release_process.md:
```diff
@@ -0,0 +1,30 @@
+
+## Tag the release
+
+- [ ] Commit the PR to the main branch
```
We should discuss this: I don't think changing the version number e.g. from 0.2.2 to 0.2.3 should wait until we are ready to release 0.2.3 - in fact, I think it's the opposite, we should update to 0.2.3 right after we release 0.2.3.

If we agree on this, the very first bullet in this list is just a verification that the version in Cargo.toml is correct.
and the very last step should be to update the version to the next release version and create a PR for it

betarelease replied:
that is correct we should update the version as soon as we have done the release. Although that is not the current state. So I will leave it that way and we will do it the way you are suggesting for 0.2.2 and update the document.

betarelease commented:
@erwin1 I updated based on your comments and addressed the overall concern in your comments instead of taking them as it is. I added some headers to clarify what we are doing at each step. Also I kept tagging as the activity of just putting the tag. Other steps I collected under a different header.

betarelease commented:
I also addressed @fishseabowl and @ihcomega56 comments. So much good feedback.

sbtaylor15 reviewed (approved):
/lgtm
